### PR TITLE
ns11: Schema for typed NICOS cache entries

### PR DESCRIPTION
**Pull request by nikhilbiyani**
- [x] If there are new schema in this PR I have added them to the list in README.md

Schema for writing cache entries from NICOS with properly typed values. Following types are supported:

- Numerical/String types
- Dict
- Tuple
- List
- Set

### CODE REVIEW

File: schemas/ns11_typed_cache_entry.fbs [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+table Dict { value: [DictMapping]; }
+table ArrayDict { value: [Dict]; }
+
+/// pylint: skip-file
+table CacheEntry {
+        key: string;            // key for this entry (usually nicos/device/parameter)
+        time: double;           // time (in seconds after epoch) when this entry was set
+        ttl: double;            // time to live (in seconds after time field of this entry)
```
Comment (matthew-d-jones) [inline schemas/ns11_typed_cache_entry.fbs]: Is this abbreviation used everywhere in NICOS? If not I'd rather use full words; I'm probably not the only one whose brain gets to "transistor-transistor logic" before "time to live".

Comment (nikhilbiyani) [reply schemas/ns11_typed_cache_entry.fbs]: Yes inside NICOS, it is the standard way to write time to live. In the other schema ns10 we used ttl as well. So, I used ttl here to be consistent everywhere. But if required I can change it to full form.

**Comment by zjttoefs** [comment]
@nikhilbiyani could you either somehow present this NICOS-Kafka-Caching in a telco or point us at some documentation? It appears to duplicate work that the forwarder is doing. But basically I don't understand how it works and I have the feeling I'm not alone.

**Comment by nikhilbiyani** [comment]
Yes, we can do the telco. Following documentation might be useful as well:

https://forge.frm2.tum.de/nicos/doc/nicos-master/services/cache/
https://forge.frm2.tum.de/nicos/doc/nicos-master/protocols/cache/

File: schemas/ns11_typed_cache_entry.fbs (hunk 1 above)
Comment (matthew-d-jones) [reply schemas/ns11_typed_cache_entry.fbs]: As this schema is very NICOS-specific I think `ttl` is fine then.

**Comment by zjttoefs** [comment]
We talked to Mark bit over a month ago. We're still hoping to get some explanation of this - not the NICOS cache in general. Looks like Kafka partially gets the same information from the NICOS cache and the EPICS forwarder. Is that correct? Is that needed?

**Comment by nikhilbiyani** [comment]
@zjttoefs. That's a valid point! We discussed this briefly in the last ECP Workshop. Yes, it looks like the data is duplicate but on a closer look you will find that in most cases it is not:

- NICOS Cache values store derived data and forwarder data is raw data:
Forwarder writes the PV updates on Kafka and can be considered as raw data that is used in many places in NICOS to derive many quantities. Each device in NICOS contain several dynamic parameters such as value, status, target, offset, user (software) limits and other static properties such as units, hardware limits, related to every device. 

In few devices such as EPICS motors, some of these parameters might directly correspond to the PV values. In these cases, yes, the data is repeated. In many cases you will also find that these parameters are derived from several PVs. As Kafka is the point of collection of all data, these parameter values need to be recorded in Kafka and NICOS cache is the best and most efficient way of doing it.

Some devices are completely derived such as virtual motors or monochromators which takes the current value of multiple motors to derive it's own position. Many devices such as slit systems use multiple motors to provide few values such as width, height. The user might not be interested in the individual motor positions, they just see and change the height and width. As NICOS exposes these parameters to users these are all written inside NICOS Cache. All these values might be varied over time during measurements and in turn tracked in NICOS Cache. Furthermore, the user might be interested in getting those values in NeXus files.

User interactions such as user commands, user logs, experiment and proposal details, sample names etc. are only recorded in NICOS. This data should also be written to Kafka.

To write all of this data inside Kafka, the best point of contact between NICOS and Kafka in NICOS Cache. Thus this solution was adopted that NICOS Cache reads/writes to Kafka instead of text files (current used in NICOS).

Although some data is redundant, it is quite small in terms of size (for e.g. the detector data is not written in NICOS Cache) and is of different kind than forwarder data.

**Comment by zjttoefs** [comment]
I fully agree that it's not all the same. User interaction is fully separate. However there are two interesting cases:

For simplicity many devices in NICOS should directly mirror the EPICS PVs, without conversion. There the forwarder offers the better updating regime with monitoring instead of the polling NICOS, I think, uses. So that you make a case for NICOS being able to accept and parse standard forwarder messages.

The other aspect is that we are need to be able to record some of the NICOS converted or generated values in the file writer. So they need to be compatible with what the file writer can deal with. More likely the file writer needs to be extended, but we should have a limited problem space.

Unfortunately, I'll be away for a bit now and I'll have to pick this up again in August.

**Pull request merged**